### ISSUE TITLE
`alt-text` removed by Quarto at least for `docx` (works with Pandoc)

### ISSUE THREAD

**Issue by mcanouil** [quarto-dev/quarto-cli#5514]
Considering the below example:

```md
## Section

![Caption for Quarto Logo](https://quarto.org/quarto.png){alt-text="Quarto Logo"}
```

|`quarto pandoc index.qmd --from markdown --to docx -o index.docx` | `quarto render index.qmd --to docx`|
|:-:|:-:|
| <img width="1624" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/8896044/ad097557-f69e-4045-9b09-f9d9d3f977d0"> | <img width="1624" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/8896044/bfd7bb63-1a78-43d5-8f19-65bf8edb0fad"> |

The initial question also mentioned `pdf`, but I don't even know how to show alternative text in PDF.

---

### Discussed in https://github.com/quarto-dev/quarto-cli/discussions/5511

<div type='discussions-op-text'>

<sup>Originally posted by **IZE85** May 12, 2023</sup>
In MS Word, one is able to "add alternative text to a shape, picture, chart, SmartArt graphic, or other object to help people with visual disabilities."

When I add alternative text to a figure via "fig-alt" it gets exported via HTML. Unfortunately, when knitting to formats like Word or PDF, my fig-alt text is saved/deposited somewhere, if I'm correct. At least, for MS Word, I cannot see the alternative text in the "alternative text" field, MS Word provides for figures.

My question: Is it possible for the fig-alt text to find its way into the MS Word Alternative text field and have something similar for PDFs?

Thank you in advance!

Ad alternative text:
"Alt text helps people with visual disabilities understand pictures and other graphical content. When someone using a screen reader comes across a picture in a document, they will hear the alt text describing the picture; without alt text, they will only know they've reached a picture without knowing what the picture shows."
Source: https://support.microsoft.com/en-us/office/add-alternative-text-to-a-shape-picture-chart-smartart-graphic-or-other-object-44989b2a-903c-4d9a-b742-6a75b451c669</div>

**Comment by cderv** [comment]
The alt text syntax with pandoc should be the one described in https://pandoc.org/MANUAL.html#images

I don't know about this `alt-text` attributes.  In your screenshort, the caption text is just reused by Pandoc as an alt text. 

Note that this works with Docx too 
````markdown
---
title: doc
format: docx
---

## Section

![Caption for Quarto Logo](https://quarto.org/quarto.png "Some alt text")
````

With Quarto we support `fig-alt` attributes,  https://quarto.org/docs/authoring/figures.html#alt-text
but it seems we are not indeed setting it for docx output the way that Pandoc's handle it. 

This would be what to fix IMO.

**Comment by cderv** [comment]
> This would be what to fix IMO.

We just don't support it in fact - we only support `fig-alt` for HTML output 
Feature would be to add in https://github.com/quarto-dev/quarto-cli/blob/59a865f351f92431eca8454124afe4c345517c2f/src/resources/filters/quarto-pre/figures.lua#L42-L50

So feature would need to be added there to leverage how Pandoc handles title text to alt text for Docx

**Comment by mcanouil** [comment]
Since computation in Quarto, uses `fig-alt`, it feels like a bug more than a missing feature, but both are fine by me.

**Comment by cderv** [comment]
Notes from quick look into this : 

* Openxml `descr` attributes it set from Caption with Pandoc, hence the Pandoc behavior we see. We remove caption in our figure handling, (and also resolve numbering in it)
* Open xml `title` can be set and appear in the docx alt text box - we could hack this to append or replace with our alt text to it but it would be set to wrong attributes in openxml

So we need some clever processing in the right order to be able to do the right things - or output directly the openxml instead of Pandoc.

**Comment by DDGerasimova** [comment]
Hi, I am wondering if there are any updates on fig-alt for docx and pdf?

**Comment by cderv** [comment]
Not yet. Issue will be updated when there is some. You should be able to subscribe to the issue to get notification - no need to reply and ask.

**Comment by DDGerasimova** [comment]
Hi - 

I am not sure if I should open a new issue, but it's related to alt text, so I thought it might fit here. The issue is about the reading order when there are pictures with alt text. If one has a title and a chart, then the reading order goes 1-chart alt text, then 2-title. However, if a title is not formatted as a title but as general text, then the reading order is correct (1-title, 2-chart alt text). [I checked for the reading order in Adobe (the accessibility tool) on a pdf file created via the 'print as pdf' option of an html file].

**Comment by mcanouil** [comment]
@DDGerasimova Could you share a small self-contained "working" (reproducible) example to work with, _i.e._, a complete Quarto document or a Git repository? Thanks.

You can share a Quarto document using the following syntax, _i.e._, using more backticks than you have in your document (usually four ` ```` `).

`````md
````qmd
---
title: "Reproducible Quarto Document"
format: html
---

This is a reproducible Quarto document using `format: html`.
It is written in Markdown and contains embedded R code.
When you run the code, it will produce a plot.

```{r}
plot(cars)
```

The end.
````
`````

**Comment by DDGerasimova** [comment]
@mcanouil, sure, I think this example below should work? Apologies that it's not very pretty, new to Quarto -- still learning :) 

The example is for the title being formatted as a title; so, to run with the title as text, I just removed #. 

````
---
 fontsize: 20pt
---

# **My test page**

```{r, include = FALSE}
library(palmerpenguins)
```

```{r, include = FALSE}
library(tidyverse)

A <- c(3, 4, 8, 10, 13)
B <- c(1, 2, 3, 4, 5)

df <- data.frame(A, B)

```

```{r, fig.alt = str_glue("Line chart illustrating the increase in A, as B increases from 1 to 5")} 
#| echo: false

library(ggplot2)

spline_A <- as.data.frame(spline(df$B, df$A))

ggplot(df) + 
  geom_point(aes(x = B, y = A), size = 2, colour = 'black') + 
  geom_line(data = spline_A, aes(x = x, y = y), colour = 'black') +
  labs(y= "A", x = "B") + theme(text = element_text(size=14))
```
````

**Comment by mcanouil** [comment]
Ok, so the same issue as the one described here.

Side note, @DDGerasimova don't mix inline code cell option and YAML style code cell option.
Quarto uses and recommends YAML.
For instance, your example stripped of unnecessary code:

````qmd
---
title: "My test page"
format: docx
---

```{r} 
#| echo: false
#| fig-alt: "Line chart illustrating the increase in A, as B increases from 1 to 5"
library(ggplot2)
A <- c(3, 4, 8, 10, 13)
B <- c(1, 2, 3, 4, 5)
df <- data.frame(A, B)
spline_A <- as.data.frame(spline(df$B, df$A))

ggplot(df) + 
  geom_point(aes(x = B, y = A), size = 2, colour = 'black') + 
  geom_line(data = spline_A, aes(x = x, y = y), colour = 'black') +
  labs(y= "A", x = "B") + theme(text = element_text(size=14))
```
````

**Comment by DDGerasimova** [comment]
I guess I meant to say 'heading' rather than 'title' -- sorry! Although with 'Title' as in your code the result is the same as with 'section' (#). So to illustrate what I meant regarding reading order: 
(The numbers 1 and 2 at the beginning of the heading and of the chart denote the order).  

(1) Examining reading order for the heading as regular text: 

<img width="960" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/93291767/b614a4f5-afcf-4f21-9fa7-28616e6ceeef">


(2) Examining reading order for the heading as section or title: 

<img width="954" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/93291767/b1911f26-0b57-4613-9f40-036e5e83bd22">

(To re-iterate: these were generated as html, then printed to pdf)

**Comment by DDGerasimova** [comment]
Regarding mixing inline code cell option and YAML style code cell option... So, my actual code is a bit more complicated than what I originally posted (please see below). My alt text includes references to the data so that actual numbers can be automatically included in the alt text. I tried to write it YAML style first but couldn't get it to work. I tried inline, and it worked. Is there a way to do this YAML style?

````
---
title: "My test page"
---

```{r}
#| include: false
library(palmerpenguins)
```

```{r}
#| include: false
library(tidyverse)

A <- c(3, 4, 8, 10, 13)
B <- c(1, 2, 3, 4, 5)

df <- data.frame(A, B)

d_start <- df[df$B == '1', ]
d_end <- df[df$B == '5', ]

```

```{r, fig.alt = str_glue("Line chart illustrating the increase in A from ", d_start %>% select(A) %>% slice(1) %>% pull, " to ", d_end %>% select(A) %>% slice(1) %>% pull, " as B increases from 1 to 5")} 
#| echo: false

library(ggplot2)

spline_A <- as.data.frame(spline(df$B, df$A))

ggplot(df) + 
  geom_point(aes(x = B, y = A), size = 2, colour = 'black') + 
  geom_line(data = spline_A, aes(x = x, y = y), colour = 'black') +
  labs(y= "A", x = "B") + theme(text = element_text(size=14))
```
````

**Comment by mcanouil** [comment]
Still don't mix. If you use inline for a code cell, use inline. If you use YAML, use only YAML.
Because at some point when mixing you'll very likely set the option twice and then wonder why your changes do not show up because only one of the option is actually used.

`knitr` syntax (it's not Quarto syntax) is:
```
#| fig-cap: !expr "1+1"
```
See <https://quarto.org/docs/computations/r.html#chunk-options>.

**Comment by DDGerasimova** [comment]
Yeah, I understand that, just can't get it to work with YAML :( When my alt text has a mix of words and code for pulling numbers from the data, I either can't get Quarto to run at all or it runs but the code for pulling numbers displays the code itself in the alt text and not the actual number...

**Comment by cderv** [comment]
Adding correct support for alt text in Quarto for docx output is not that straightfoward. I moved this to 1.5

**Comment by ARMurray** [comment]
I'd like to add powerpoint (.pptx) as a needed component of this fix. I assume that support for alt text to word will extend to powerpoint but it has not been mentioned so far.

**Comment by srvanderplas** [comment]
I'd just like to bump @ARMurray's issue -- this is particularly important for any public entities using quarto to generate PPTX files, because the interpretation of the ADA requires WCAG2.0 AA-level compliance by April 2026. I know that it's a pain in the rear to implement, but it will affect the ability of university and government users to leverage quarto.

**Comment by cscheid** [comment]
I want to make sure we're being extremely clear here, because there's enough confusion as is.

> I know that it's a pain in the rear to implement,

I hear you, but let me assure you that the issue isn't just that it's a pain in the ass - it's that we are fundamentally limited in what we can do with .pptx given the way that pptx writing works in Pandoc.

> using quarto to generate PPTX files, because the interpretation of the ADA requires WCAG2.0 AA-level compliance by April 2026

There must be some confusion here somewhere, because PPTX is a format from Microsoft Office. WCAG stands for "Web Content Accessibility Guidelines", emphasis on "Web Content". I don't understand how we can draw conclusions from WCAG requirements to non-web formats like PPTX.

**Comment by srvanderplas** [comment]
Yeah, sorry, that's on me. My university bungled the messaging a bit around
title ii and wcag, but it's an issue that many of us are facing.

Title 2 requires accessible content, whether web or not. Since pptx has the
capacity to represent alt text, it should be preserved in the markdown
-pandoc-pptx conversion.

Given that alt text  is preserved for images but not computations, it seems
like it would be possible to make it work for computations as well?

Markdown is actually a very efficient way to generate accessible content
(as I'm constantly telling colleagues) in that it focuses on informative
content with only a minimum of (meaningful) formatting, but that only helps
if the extra information for accessibility is preserved in each output
format.

**Comment by cscheid** [comment]
> Title 2 requires accessible content, whether web or not. Since pptx has the
capacity to represent alt text, it should be preserved in the markdown
-pandoc-pptx conversion.

I hear you.

But in order to really fix this, we would need more information about what "requires accessible content" means, and more specifically how they're going to check for PPTX. For WCAG, we have (and use in 1.8) axe-core. But I don't know what tooling or what expectations are there for PPTX.

In addition, Pandoc's pptx writer works very differently from the docx writer, and Quarto is fundamentally limited by what we can get from Pandoc in this case.

> Given that alt text  is preserved for images but not computations, it seems
like it would be possible to make it work for computations as well?

I'm confused. I think this question is related to the docx issue, but I'm only responding to the pptx vs wcag concerns here.

In fact, now that I think about this, we should probably move the pptx discussion to its own issue.

**Comment by cscheid** [comment]
For reference:

- #12956
- #12957

**Comment by bfisseler** [comment]
> There must be some confusion here somewhere, because PPTX is a format from Microsoft Office. WCAG stands for "Web Content Accessibility Guidelines", emphasis on "Web Content". I don't understand how we can draw conclusions from WCAG requirements to non-web formats like PPTX.

Sorry for answering to a nearly one year old message but as I am currently researching Quarto and accessibility, I also came across this issue and the corresponding [parent issue 14348](https://github.com/quarto-dev/quarto-cli/issues/14348).

It is correct that WCAG 2 is the abbreviation for "Web Content Accessibility Guidelines". But this will change in the future and has mainly historical reasons. The upcoming WCAG 3 will be technology agnostic and be the abbreviation for ["W3C Accessibility Guidelines"](https://www.w3.org/WAI/standards-guidelines/wcag/wcag3-intro/). 

The European Norm EN 301 549 refers to WCAG 2.1 and already requires non-web documents to be fully accessible. E.g. requirement 10.1.1.1 states that [all non-web documents shall satisfy WCAG success criterion 1.1.1](https://www.etsi.org/deliver/etsi_en/301500_301599/301549/03.02.01_60/en_301549v030201p.pdf) aka alt-text (and other alternatives) for non-text content. Additional checks apply such as sufficient contrast, use of slide layouts (pptx) and built-in styles (DOCX) etc.

The way-to-go check currently is the [accessibility checker integrated in the Microsoft Office](https://support.microsoft.com/en-us/office/improve-accessibility-with-the-accessibility-checker-a16f6de0-2f39-4a2b-8bd8-5ad801426c7f) apps. A few websites also offer "free" accessibility checks of DOCX and PPTX documents but I have not tested these thoroughly.
